utils/vector.hpp

```diff
@@ -0,0 +1,366 @@
+unsigned int Vector<T>::getCapacity() const
+{
+    return capacity;
+}
+
+
+template<class T>
+void Vector<T>::resize(unsigned int newSize)
+{
+    if(newSize < size)
+        size = newSize;
+    else if(newSize > size)
+    {
+        //Reallocate the array
+        T* newArray = new T[newSize];
+
+        //Copy the values of the existing array to new array
+        for(unsigned int i = 0; i < size; i++)
+        {
+           newArray[i] = array[i];
+        }
+
+        //Change the data of the vector
+        delete[]   array;
+        array    = newArray;
+        capacity = newSize;
+        size     = newSize;
+    }
+}
+
+
+template<class T>
+void Vector<T>::resize(unsigned int newSize, const T& value)
+{
+    if(size > newSize)
+        size = newSize;
+    else if(size < newSize)
+    {
+        //Reallocate the array
+        T* newArray = new T[newSize];
+
+        //Copy the values of the existing array to new array
+        for(unsigned int i = 0; i < size; i++)
+        {
+            newArray[i] = array[i];
+        }
+
+        //The left over elements will contain the specified value
+        for(unsigned int i = size; i < newSize; i++)
+        {
+            newArray[i] = value;
+        }
+
+        //Change the data of the vector
+        delete[] array;
+        array    = newArray;
+        capacity = newSize;
+        size     = newSize;
+    }
+}
+
+
+template<class T>
+void Vector<T>::reserve(unsigned int newCap)
+{
+    if(capacity < newCap)
+    {
+        //Keep multiplying capacity by 2 until it's larger than the new capacity
+        while(capacity < newCap)
+            capacity *= 2;
+
+        //Reallocate the array
+        T* newArray = new T[capacity];
+
+        //Copy the values of the existing array to new array
+        for(unsigned int i = 0; i < size; i++)
+        {
+            newArray[i] = array[i];
+        }
+
+        //Delete old array and assign new array to vector
+        delete[] array;
+        array = newArray;
+    }
+}
+
+
+
+/* Modifiers */
+template<class T>
+void Vector<T>::push_back(const T& value)
+{
+    if(capacity == size)
+    {
+     resize(size + 1);
+    }
+    array[size] = value;
+}
+
+
+template<class T>
+void Vector<T>::pop_back()
+{
+    if(size > 0)
+    {
+        size--;
+    }
+}
+
+
+template<class T>
+void Vector<T>::clear()
+{
+    size = 0;
+    capacity = 0;
+    delete[] array;
+    array = nullptr;
+}
+
+
+
+/* Operator Overloads */
+template<class T>
+Vector<T>& Vector<T>::operator=(const Vector& vec)
+{
+    //Delete old array
+    delete[] array;
+
+    //Resize array to fit new elements
+    resize(vec.size);
+
+    //Deep copy elements from other vector
+    for (unsigned int i = 0; i < size; i++)
+    {
+        array[i] = vec.array[i];
+    }
+
+    return *this;
+}
+
+
+template<class T>
+Vector<T>& Vector<T>::operator=(Vector&& vec)
+{
+    size     = vec.size;
+    array    = vec.array;
+    capacity = vec.capacity;
+    vec.array= nullptr;
```

Comment:
No spacing between operators and variables.
```suggestion
    vec.array = nullptr;
```
This will throw off the lining above it, so make sure they line up if you really want them to line up.